src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.AddProperty.cs

```diff
@@ -209,18 +210,26 @@ internal JsonPropertyInfo CreateRootObject(JsonSerializerOptions options)
                 options: options);
         }
 
-        internal JsonPropertyInfo CreatePolymorphicProperty(JsonPropertyInfo property, Type runtimePropertyType, JsonSerializerOptions options)
+        internal JsonPropertyInfo GetOrAddPolymorphicProperty(JsonPropertyInfo property, Type runtimePropertyType, JsonSerializerOptions options)
         {
-            JsonPropertyInfo runtimeProperty = CreateProperty(
-                property.DeclaredPropertyType, runtimePropertyType,
-                property.ImplementedPropertyType,
-                property.PropertyInfo,
-                parentClassType: Type,
-                converter: null,
-                options: options);
-            property.CopyRuntimeSettingsTo(runtimeProperty);
+            static JsonPropertyInfo CreateRuntimeProperty(Type runtimePropertyType, (JsonPropertyInfo property, JsonSerializerOptions options, Type classType) arg)
+            {
+                JsonPropertyInfo runtimeProperty = CreateProperty(
+                    arg.property.DeclaredPropertyType, runtimePropertyType,
+                    arg.property.ImplementedPropertyType,
+                    arg.property.PropertyInfo,
+                    parentClassType: arg.classType,
+                    converter: null,
+                    options: arg.options);
+
+                arg.property.CopyRuntimeSettingsTo(runtimeProperty);
+
+                return runtimeProperty;
+            }
+
+            var cache = LazyInitializer.EnsureInitialized(ref property.RuntimePropertyCache, () => new ConcurrentDictionary<Type, JsonPropertyInfo>());
```

Comment:
Is this preferred over `Lazy<T>`?

Reply:
I think it's preferred here, since `Lazy<T>` would require an allocation of an instance of `Lazy<T>` for every instance of `JsonPropertyInfo` (even if it didn't have any polymorphic properties).

But that's a moot point anyway if we're going to reuse `JsonClassInfo.PropertyCache` (more on that later).